File: ncurses/plantinfo.cpp
#include "plantinfo.h"

#include "visettings.h"
#include "legend.h"
#include <stdio.h>
#include <string.h>
#include <cstdlib>
#include <fstream>


Plantinfo::Plantinfo( int x, int y, int width, int height )
  : Window( x, y, width, height )
{


}

#define INFO_PRINT( y, z ) { mvwprintw( mWindow, i++, 2, #y": %d        ", getAttr(plant)->z ); }

void Plantinfo::update()
{

  werase(mWindow);
  box( mWindow, 0, 0 );
  wattron( mWindow, COLOR_PAIR( DEFAULT_COLOR ) );


  if( getAttr(plant) )
  {

    int i = 3;
    if( !getAttr( autoEnd ) )
    {
      mvwprintw( mWindow, 1, 2, "Plant Info" );


      INFO_PRINT( Health, health );
      INFO_PRINT( Root Level, rootLevel );
      INFO_PRINT( Leaf Level, leafLevel );
      INFO_PRINT( Flower Level, flowerLevel );
      INFO_PRINT( Root LvlUp, rootLevelUp );
      INFO_PRINT( Leaf LvlUp, leafLevelUp );
      INFO_PRINT( Flower LvlUp, flowerLevelUp );
    } else {

      std::system( "curl -s 131.151.189.189/api/top > topSeeds " );
      std::ifstream in( "topSeeds", ios::in );
      std::string seed, seeds[3];

      in >> seed;

      int k = 0;
      int start = 0;
      for( int i = 0; i < seed.length(); i++ )
      {
        if( seed[i] == ',' || i == seed.length()-1 )
        {
          seeds[k++] = seed.substr( start, i-start );
          start = i+1;
          if( k == 3 )
            break;
        }
      }

      seeds[2] += seed[seed.length()-1];

      mvwprintw( mWindow, 1, 2, "Top Seeds (pun lol)" );
      mvwprintw( mWindow, i++, 2, "1) %s", seeds[0].c_str() );
      mvwprintw( mWindow, i++, 2, "2) %s", seeds[1].c_str() );
      mvwprintw( mWindow, i++, 2, "3) %s", seeds[2].c_str() );


    }
    // Check if plant is talking
    GameState state = *getAttr( state );
    ++i;
    for( std::vector<Animation *>::iterator j = state.animations.begin(); j != state.animations.end(); j++ )
    {
      if( (*j)->type == TALK )
      {
        if( getAttr( plant )->objectID == ((Talk *)(*j))->plantID )
        {

          char *msg = ((Talk *)(*j))->message;
          bool nLine = false;


          for( int l = 0; l < 28; l++ )
          {
            if( msg[l] == '\n' )
            {
              nLine = true;
            }
          }

          while( strlen( msg ) > 28 || nLine )
          {
            nLine = false;
            char str[29];
            strncpy( str, msg, 28 );
            int n = 28;

            for( int l = 0; l < 28 && l < strlen(msg); l++ )
            {
              if( str[l] == '\n' )
              {

                str[l] = '\0';
                n = l;
                break;
              }

            }
            str[28] = '\0';
            mvwprintw( mWindow, i++, 2, "%s", str );
            msg += n+1;

            for( int l = 0; l < 28; l++ )
            {
              if( msg[l] == '\n' )
              {
                nLine = true;
                break;
              }
            }

          }
          mvwprintw( mWindow, i++, 2, "%s", msg );
        }
      }
    }


  } else {
    if( !getAttr( autoEnd ) )
    {
      mvwprintw( mWindow, 1, 2, "Empty" );
    }
    else
    {
      //
    }
  }

  box( mWindow, 0, 0 );
  wrefresh( mWindow );

}

File: pycurses/plantinfo.py
# -*- coding: utf-8 -*-
"""
Created on Wed Apr 14 10:00:30 2010

@author: jacob
"""

import curses
import libparsermodule as pm
from legend import Legend
from visettings import Visettings
from window import Window


class Plantinfo(Window):
    def __init__(self, x, y, width, height ):
        Window.__init__(self, x, y, width, height )
    
    def printInfo(self, n, str):
        self.Window.addstr( n, 2, str + "      " )
        n+=1
        return n

    def update(self):
        self.Window.erase()
        self.Window.box(0,0 )
        self.Window.attron( curses.color_pair( Legend.DEFAULT_COLOR ) )
        
        if Visettings.Plant:
            i = 3
            if not Visettings.AutoEnd:
                self.Window.addstr( 1, 2, "Plant Info" )
                info = ["Health: %d" % (Visettings.Plant.health ),
                "Roof Level: %d" % (Visettings.Plant.rootLevel ),
                "Leaf Level: %d" % (Visettings.Plant.leafLevel ),
                "Flower Level: %d" % (Visettings.Plant.flowerLevel ),
                "Root LvlUp: %d" % (Visettings.Plant.rootLevelUp ),
                "Leaf LvlUp: %d" % (Visettings.Plant.leafLevelUp ),
                "Flower LvlUp: %d" % (Visettings.Plant.flowerLevelUp )
                ]


                for k in info:
                    i = self.printInfo( i, k )
                # Print Stuff
            else:
                pass
                # LIB CURL TOP 3 SEEDS
                
            # TALKING: Need to fix this error
            for j in pm.getAnims(Visettings.State):
                if j.type == 7:
                    if Visettings.Plant.objectID == j.plantID:
                        w = 0
                        msg = pm.getTalk( j )
                        nLine = 0

                        for k in range( 0, 28 ):
                            if msg[k] == '\n':
                                nLine = 1

                        l = 0

                        while len( msg )-w > 28 or nLine:
                            l = w+28
                            nLine = 0

                            for k in range( w, l ):
                                if msg[k] == '\n':
                                    l = k
                                    break                          
                            self.Window.addstr( i, 2, msg[w:l] )
                            w = l+1
                            i+=1

                            for k in range( w, w+28 ):
                                if k < len(msg) and msg[k] == '\n':
                                    nLine = 1
                                    break
        
                        self.Window.addstr( i, 2, msg[w:] )
                        i+=1
        else:
            if not Visettings.AutoEnd:
                self.Window.addstr( 1, 2, "Empty" )
            else:
                pass
        
        self.Window.box( 0, 0 )
        self.Window.refresh()
